### ISSUE TITLE
Assigning service published ports to IP

### ISSUE THREAD

**Issue by rtuk** [docker/docker#26696]
**Description**
When you create a service with a published port, this port is assigned to all host IPs (0.0.0.0). You cannot create a new service on the same port because this port is already in use.
This issue was first created in [docker/swarm](https://github.com/docker/swarm/issues/2478)

**Steps to reproduce the issue:**
1. docker service create --name test -p 192.169.2.8:80:80 nginx

and
1. docker service create --name test -p 80:80 nginx
2. docker service create --name test2 -p 80:80 nginx

**Describe the results you received:**
invalid argument "192.169.2.8:80:80" for p: HostIP is not supported by a service.
See 'docker service create --help'.

and
1. 1wwed5dix1xga27583dh9bh6r
2. Error response from daemon: rpc error: code = 3 desc = port '80' is already in use by service 'test' (1wwed5dix1xga27583dh9bh6r)

**Describe the results you expected:**
If a host has multiple IP addresses it should be possible to publish container ports on a specific IP address.

**Additional information you deem important (e.g. issue happens only occasionally):**

**Output of `docker version`:**

```
Client:
 Version:      1.12.1
 API version:  1.24
 Go version:   go1.7.1
 Git commit:   6f9534c
 Built:        Thu Sep  8 10:31:18 2016
 OS/Arch:      darwin/amd64

Server:
 Version:      1.12.1
 API version:  1.24
 Go version:   go1.6.3
 Git commit:   23cf638
 Built:        Thu Aug 18 17:52:38 2016
 OS/Arch:      linux/amd64
```

**Output of `docker info`:**

```
Containers: 0
 Running: 0
 Paused: 0
 Stopped: 0
Images: 0
Server Version: 1.12.1
Storage Driver: aufs
 Root Dir: /var/lib/docker/aufs
 Backing Filesystem: extfs
 Dirs: 1
 Dirperm1 Supported: true
Logging Driver: json-file
Cgroup Driver: cgroupfs
Plugins:
 Volume: local
 Network: null bridge overlay host
Swarm: active
 NodeID: 14agfq1cb2asswaxlj9m9m4qd
 Is Manager: true
 ClusterID: dcj2b2yl0lt7k0b1wyq2lfl3j
 Managers: 1
 Nodes: 1
 Orchestration:
  Task History Retention Limit: 5
 Raft:
  Snapshot Interval: 10000
  Heartbeat Tick: 1
  Election Tick: 3
 Dispatcher:
  Heartbeat Period: 5 seconds
 CA Configuration:
  Expiry Duration: 3 months
 Node Address: 192.168.65.2
Runtimes: runc
Default Runtime: runc
Security Options: seccomp
Kernel Version: 4.4.20-moby
Operating System: Alpine Linux v3.4
OSType: linux
Architecture: x86_64
CPUs: 4
Total Memory: 1.952 GiB
Name: moby
ID: BLI4:JSAG:E5EJ:YI67:QCOY:5PVH:JOZA:TJEQ:IX7Q:A7CT:YYXQ:GE6T
Docker Root Dir: /var/lib/docker
Debug Mode (client): false
Debug Mode (server): true
 File Descriptors: 34
 Goroutines: 120
 System Time: 2016-09-19T08:05:04.123825143Z
 EventsListeners: 1
No Proxy: *.local, 169.254/16
Registry: https://index.docker.io/v1/
Insecure Registries:
 127.0.0.0/8
```

**Additional environment details (AWS, VirtualBox, physical, etc.):**
Native docker on Mac OSX El Capitan (10.11.6)

**Comment by thaJeztah** [comment]
I don't think this is really possible. In a multi-node swarm setup (which would be the "regular" setup for swarm mode), you'd have to know and specify the IP-address of *every* node in the swarm. I know there's been talk about "federation" (manage groups of nodes / multiple swarms), which is "slightly" related, but wouldn't cover this use case.

ping @mrjana any thoughts?

**Comment by evanp** [comment]
I think the work-around for this is to publish different ports for the two services (like 8001 and 8002) and then use a load balancer bound to each IP address to proxy to each port.

Unfortunately the load balancer has to be managed outside of the cluster. They can either be managed through a Docker container on another virtual machine, or they can be managed outside of Docker. Many cloud platforms provide a load balancer service separate from the virtual machine service, though.

**Comment by thaJeztah** [comment]
For web-services, using a proxy-service that directs requests to the right service, based on hostname is the usual approach, also see this repo for a nice writeup; http://github.com/stevvooe/sillyproxy

**Comment by mrjana** [comment]
Yes, this is a features that was requested a few times before. We have given a bit of thought but as @thaJeztah said all of these service creations happen at cluster level. Nevertheless, there is a possibility of somehow selecting an IP address on any host based on some selection criteria specified at service level. Even though this leads to more flexibility to users, it does lead to a lot of complexity in docker. But if there are enough interests on this feature,  we may still add this in the future.

**Comment by michelvocks** [comment]
> This issue was first created in docker/swarm

Just as a small remark. With docker/swarm this is possible by exposing the port and also the interface ip address like described here: https://docs.docker.com/engine/reference/run/#/expose-incoming-ports
An example can be seen [in this comment.](https://github.com/docker/swarm/issues/2478#issuecomment-248084147)

@mrjana Don't you think something similar should be possible with swarmkit?

**Comment by aluzzardi** [comment]
This is working as intended

> I think the work-around for this is to publish different ports for the two services (like 8001 and 8002) and then use a load balancer bound to each IP address to proxy to each port.

Correct

> Unfortunately the load balancer has to be managed outside of the cluster. They can either be managed through a Docker container on another virtual machine, or they can be managed outside of Docker. Many cloud platforms provide a load balancer service separate from the virtual machine service, though.

Not really - your load balancer could be manager by Swarm. For instance, you could deploy an `nginx` on port 80 that redirects traffic to other services.

Does that answer your question?

**Comment by rtuk** [comment]
@mrjana implementing this feature would be really appreciated, we are looking forward to updates on the subject.
@aluzzardi the concept is clear, we will be implementing a load balancer to solve our problem, and hopefully changing it to assigned IP addresses if and when that possibility is added.

**Comment by omallo** [comment]
I also think that such a feature would be interesting, not only for cases like the one outlined by @rtuk , but also for scenarios in which you want to publish a service port to your private network so that other tools *outside* Docker can interact with the service but you do not want to expose the port on your public network.

In order to support such a feature on service level, one possibility might be to specify a network mask which restricts the network interfaces a published port binds to when a service task is executed on a specific node (of course, the network interface might not always be unique).

**Comment by aluzzardi** [comment]
> 1. docker service create --name test -p 192.169.2.8:80:80 nginx

When you are doing this, what are you expecting exactly? /cc @YouriT @rtuk 

Are you expecting for the service to be scheduled on a machine that has that IP address available?

**Comment by aluzzardi** [comment]
Maybe one solution would be a per-node option to specify which interface/IPs ingress should listen on

**Comment by YouriT** [comment]
@aluzzardi

## tl;dr

The choice of where (node) you schedule containers should be left to the user with the actual behaviour that Swarm Mode and it's ingress network have brought to the table. Being able to `--constraint` and using different `--mode` to schedule a service should remain untouched. 

## Longer

You should be able to attach a container to an IP address+port and it doesn't matter which node has this IP, neither whether there is even a node's interface that has the IP or not, the user should attach an IP address that he is able to handle on a node.

When traffic will reach the node that indeed has this IP, ingress should handle the rest and redirect (or not) the traffic to the right node that has the container (or straightly to the container if it's on the current node).

I think the best would be to keep the architecture of the network, ingress and overlay network as it is now. Example of configuration as I would see it:
HostA (1st node) has 2 interfaces: eth0 (public) & eth1 (private) with multiple IP addresses as following:
```
eth0
1.1.1.1/24
1.1.1.2/24
...
1.1.1.127/24

eth1
10.0.0.1/20

eth1:0
172.16.0.1/20
```

HostB (2nd node) has 2 interfaces: eth0 (public) & eth1 (private) with multiple IP addresses as following:
```
eth0
1.1.1.128/24
1.1.1.129/24
...
1.1.1.254/24

eth1
10.0.0.2/20

eth1:0
172.16.0.2/20
```

Let's say we have 2 services (ser1 & ser2) that both need to listen on port 80 but they use different protocols that are both TCP based but not the same, so one service cannot handle the protocol of the other. The easy solution is to stick each service on a different IP, ser1 would be on 1.1.1.1:80 & ser2 would be on 1.1.1.10:80.

Now let's run `docker service create -p 1.1.1.1:80:80 --replicas 10 ser1` and `docker service create -p 1.1.1.10:80:80 --replicas 10 ser2`. Swarm would schedule 5 ser1 on HostA, 5 ser1 on HostB, 5 ser2 on HostA and 5 ser2 on HostB.

If I call 1.1.1.1:80, it would always reach HostA of course and 1.1.1.10:80 would always reach HostA too. But let's say that all nodes on HostA are busy, then I would expect every new connections to reach HostB through the overlay.

Like this, we can keep the same architecture of routing without adding any constraint to the user (he still can use all the `--mode`, and can still constrain where the containers will be scheduled too) by enabling all the flexibility of multiple IP addresses.

Hope this helps. I can refine if it's not clear enough.

**Comment by aluzzardi** [comment]
Thanks @YouriT 

Wouldn't this better addressed by using an external load balancer (e.g. ELB if you run on AWS)?

For instance, `service1` would expose port `30000` and `service2` port `30001`.

Then you can provision two ELBs (each with its own IP address), and one of them redirects port `80` to `30000` while the other one would redirect port `80` to `30001`.

There's an alternative solution: If you're not running on cloud (so no external load balancer), you could make your own quite easily by deploying an haproxy (or nginx) to a node or set of nodes and using `host ports` rather than ingress (to actually bind port 80 on the local machine rather than ingress).

Makes sense?

**Comment by YouriT** [comment]
@aluzzardi the problem is that you are expecting here that it's only HTTP related, for which exists plenty of proxies, load balancers etc. But if you go for some private protocols or SSH for example then you are stuck, you either need to implement one yourself or find a workaround. With the proposition above you have the flexibility of Docker in Swarm Mode coupled to the Swarm management power which follows what is pursued with Swarm Mode no?

**Comment by nickygerritsen** [comment]
For us something like this would also be beneficial.

We have an internal network and an external network and we would like to bind some services only onn the internal network (and others on the external one) for things not using Docker.

For example we might still have a legacy application that wants to use MySQL, which is running in the Docker Swarm. We can then expose port 3306 only internally.

The same holds with HTTP: we have some services that need to use HTTP API's that we want to run in Swarm. But the consumer of this API is not in Docker, so we want to expose some things on port 80/443 internally, but at the same time also expose things on 80/443 on the external network for the "outside world". Although we could use a loadbalancer for HTTP, this would result in a single point of failure, while it would be way nicer if we could just use the ingress network with Docker's load balancer feature directly.

I think we can solve this for now using some `iptables` rules (blocking some external network ports and forwarding some others to a different internal port), but this is not ideal.

**Comment by cghislai** [comment]
This is a desired feature for us as well. 
AFAIC, when publishing service port in host mode (vs ingress mode), it makes sense to be able to specify ip adresses because you probably constrained the service task to some node already. Some protocols require you to use specific ports (im thinking netbios). Altough it should be possible to proxify netbios services, it makes much sense to keep them bound to port 139 but restrict them to a specific interface.

**Comment by asbjornenge** [comment]
Just my two cents; I really need this 😛  Typically for a few entry points to my swarm where I want to control specifics like this. I'm using a custom SDN (software defined network) and custom routing (not the swarm routing mesh) for some of my services. Not a problem for me to manage constraints "manually" to avoid conflict.

**Comment by MichaelErmer** [comment]
@cghislai we have a "docker-nginx-controller" which runs in global mode, it mounts the docker socket  and starts a "local" docker container containing bound to the desired ip. You can use a trap launch script or run the nginx in interactive mode, so it gets killed when the controlled is stoped/removed/updated.

**Comment by sandys** [comment]
we would want this as well. interestingly this works in compose file perfectly - wanted to understand if that was an accidental feature (soon to be deprecated), or will you make the cli consistent with the compose file.

**Comment by thaJeztah** [comment]
@sandys for docker-compose standalone, it works, because you'll be deploying containers, not swarm/kubernetes services. For `docker stack deploy`, it depends on the version of the CLI you're using; older versions of the CLI would silently ignore the option (see https://github.com/docker/cli/pull/1017, which added a warning)

**Comment by richard-delorenzi** [comment]
First sorry if my swarm terminology is a bit wrong, I have not used it much, as this issue is a road block for me. 

I will out line some use-cases, and some suggestions.

 1. As a developer I would like to run several instances, on my local machine all on the same port. Using just docker I can map a port to 127.0.0.2:1234 for one container, and to 127.0.0.3:1234 for another. If I use swarm, then this all breaks; swarm binds to every address, and refuses to start, if something is bound to at least one address. It would be nice if the 127.0.0.0/8 addresses could be used. It may be tricky to find where the node is in the swarm, but constraints, and/or probing the configuration can be used to find it.
 2. Some times I have externally and internally facing interfaces/nodes. I would like to constrain some services to only bind to internal interfaces. We could use a list, or masks, or just specify one address. Constraints could be used to ensure that a sensitive service is not on an externally visible node.
3. Sometimes I may not even care about the address and the service being on the same node. I could implement this by having a container constrained to a node, that forwards data to the service. Then it just uses use case 2.

**Comment by cghislai** [comment]
If you have no need for the swarm ingress network, there is a workaround by using the `host` network.
Just declare the `host` network in your swarm service definition and make sure you don't declare any port, then the dockerized process will be able to bind the host interfaces the way it wants.
You will not be able to connect to the other containers through the ingress network though, so you will have to use an exposed port;

**Comment by ienovytskyi** [comment]
https://github.com/moby/moby/issues/32299#issuecomment-501397573

**Comment by TuKDoan** [comment]
+1 for this feature request (?).

Currently attempting to run a udp nginx load balancer to make multiple dns servers highly available from one ip. 

The containers are man-in-the-middle dns forwarders that intercept requests to see if they're blocked or not, think dns filter, and they have to use the host to forward dns queries to the local dns.

Each node is listening on the respective loopback ip port 53/udp to make these resolutions.

Attempting to use swarm's ingress network to bind any external client request to 53/udp to bind to 5353; which the nginx udp lb is listening on, then proxying that to port 6050/udp to the worker nodes; then 6050/udp back to 53/udp for the dns filter containers.

Client DNS Query 53 > *:53:5353/udp NGINX Load balancer <upstream server>:5353:6050 > *:6050:53 Worker > *:53 DNS filter

With `docker run` I am able to specify the ip of the NIC so it's able to receive dns queries from external clients, process through dns filter, then use the loopback:53 to resolve and forward the request to another local dns. 

`docker container run --name dnsfilter -d -p 10.0.187.84:53:53/udp -p 10.0.187.84:53:53/tcp localhost:5000/dnsfilter`

`docker service` on the other hand will just override the loopback listener causing the dns filters to not even resolve. 

`docker service create --name dnsfilter -p 6050:53/udp -p 6050:53/tcp --constraint node.role!=manager --mode global localhost:5000/dnsfilter`

`docker service create --name nginxproxy -p 53:5353/udp -p 53:5353/tcp --constraint node.role==manager -mode global localhost:5000/nginx-proxy`

I'd like to use service so I could automate adding new nodes and automate the load 
balancer configuration by making all the nodes container aware.

It works when running each container individually and listening on a specific NIC, but that proves to be cumbersome when attempting to scale.

Please see attached diagram for clarity. I'm game for any ideas or work around! 

<img width="1274" alt="Screen Shot 2019-12-20 at 11 32 50 PM" src="https://user-images.githubusercontent.com/32083324/71303632-1ea36700-2381-11ea-938a-474462ef5669.png">

**Comment by DrPyser** [comment]
So after three years, what is the right way to allow secure access to admin APIs for swarm services? 

We are in the process of migrating our docker/docker-compose infrastructure to docker swarm, and this is one significant obstacle. I have some services that expose ports for admin APIs, which I would rather not have to expose to the internet(they are currently access through ssh tunneling).

**Comment by sg-jorrit** [comment]
Another valid use-case is the ability to run service ingests on (virtual) loopback-addresses (like used with eBGP load balancing or CARP/VRRP). Introducing yet another proxy in between (HTTP only) or socat'ting/passing through traffic  seems very unnecessary and only introduces another component that could possibly fail.

Maybe restrict it to simply refusing to expose the service on a node that does not have the requested IP available, though even this can be solved with allowing binding to non-existing IP's with sysctl in most cases.

Please re-consider making it possible to bind only to specified IP's.

**Comment by jairov4** [comment]
I have other use case: it is about to have services with custom placement constraint that will ensure the task will be in right node. Then this node hold multiple NICs and I want to bind specifc NICs to specific services in `host` mode.

**Comment by tcurdt** [comment]
I am flabbergasted not just about the omission but also about questioning the usefulness.

I think for most people it's not about specifying the IP (inside the yaml's) which would indeed be problematic in a cluster setup, but about defining the interface or network segment to bind to. That's just good practice - and not problematic in a cluster setup at all.

So there is still no official support for this? and no plans?
And it also sounds like there is also no real work around. (Anyone?)

I was hoping that running `dockerd` with `--ip` would work. This would still make sense in a cluster.
And from the documentation it sounds like it should work

> --ip ip Default IP when binding container ports (default 0.0.0.0)

but that's apparently not the case.

**Comment by bderubinat** [comment]
Putting the question otherwise.

- We have a docker swarm with several nodes and several services.
- Some services are meant to be public (eg: a public web site), and some are meant to be private (eg a private web site).
- The services mean to be private are sharing their http service on a private-net docker overlay network
- The services meant to be public are sharing their http service on a public-net docker overlay network

From their I'm stuck because I don't want isolation between public and private web sites.

I want public/private segregation apply further. What I would like is to have two Nginx reverse proxies.
- One Nginx connected on docker private-net overlay network and exposing port 80 on eth1 (private network)
- One Nginx connected on docker public-net overlay network and exposing port 80 on eth0 (DMZ)

a) no, I don't want...

- these 2 Nginx to expose ports 80 and 81 and have a load balancer map INTERNET-IP:80 to port 81 while the port 80 remain untouched.
- 2 docker swarms, one public, one private

b) how do I do this? What are the options?

**Comment by smcpeck** [comment]
I've a swarm mode server with multiple public IP addresses.

On one IP address, I am able to SSH into the host.
On another, I'd like to setup GitLab in docker and allow users to do git operations over SSH on a different IP address.

This feels like a very basic use case, but since docker always wants to bind GitLab's port 22 to every IP address, things choke.

It seems like that behavior is tied to this.

It shouldn't be difficult to tell docker's ingress network to not listen/bind on 0.0.0.0.  Why can't we configure this yet?  It is quite crippling.

**Comment by thaJeztah** [comment]
There's ways to create a custom [ingress network](https://docs.docker.com/network/overlay/#customize-the-default-ingress-network), and to customise the [`docker_gwbridge` interface](https://docs.docker.com/network/overlay/#customize-the-docker_gwbridge-interface), which may help for some scenarios.

**Comment by P1X3** [comment]
I am running into same issue and not sure if I am just not following the docs to set it up properly or if this is still being tracked.
Initially this issue was observed in cloud environment, but I am able to replicate with my lab environment as well.

Node 1:
  interface enp2s0
    IP 192.168.0.10/23
    IP 192.168.0.11/23
    IP 192.168.0.14/23
  interface enp2s1
    IP 192.168.0.20/23
    IP 192.168.0.21/23

Node 2:
  interface enp2s0
    IP 192.168.0.12/23
    IP 192.168.0.15/23

Node 3:
  interface enp2s0
    IP 192.168.0.13/23
    IP 192.168.0.16/23
  interface enp2s1
    IP 192.168.0.22/23
    IP 192.168.0.23/23

Creating nginx service will listen on all IPs across all hosts and all interfaces. 
Creating same service with port published to host and constrained to one host does listen on all IPs on that specific host and its all interfaces

I do not see how creating a custom ingress network and customizing docker_gwbridge interface allows to specify the IPs or interfaces.

This limits the usage of other IPs or interfaces on the node if docker service for nginx is taking port 80 on all IPs. Running another process on the node running separately from docker trying to listen on specific IP and port 80 is not possible.

**Comment by erjoalgo** [comment]
I'm attempting to migrate a docker compose project to swarm and I'm running into this limitation. I'd like to keep my swarm cluster in a separate private network or VPN and allow a host-level reverse proxy to handle SSL and route traffic based on server names. I don't understand why swarm insists on binding on every physical interface, which makes it difficult to isolate my unencrypted application traffic on my public-facing host.
